<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Classification using Microbiome | Microbiome in R</title>
  <meta name="description" content="Chapter 8 Classification using Microbiome | Microbiome in R" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Classification using Microbiome | Microbiome in R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Classification using Microbiome | Microbiome in R" />
  
  
  

<meta name="author" content="Nabiilah Ardini Fauziyyah" />


<meta name="date" content="2020-06-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="microbiome-analysis.html"/>
<link rel="next" href="closure.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Forewords</a></li>
<li class="chapter" data-level="2" data-path="microbiome-and-beyond.html"><a href="microbiome-and-beyond.html"><i class="fa fa-check"></i><b>2</b> Microbiome and Beyond</a></li>
<li class="chapter" data-level="3" data-path="microbiome-analysis-workflow.html"><a href="microbiome-analysis-workflow.html"><i class="fa fa-check"></i><b>3</b> Microbiome Analysis Workflow</a><ul>
<li class="chapter" data-level="3.1" data-path="microbiome-analysis-workflow.html"><a href="microbiome-analysis-workflow.html#the-lab-work"><i class="fa fa-check"></i><b>3.1</b> The Lab Work</a></li>
<li class="chapter" data-level="3.2" data-path="microbiome-analysis-workflow.html"><a href="microbiome-analysis-workflow.html#bioinformatic-analysis"><i class="fa fa-check"></i><b>3.2</b> Bioinformatic Analysis</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="bioconductor-installation.html"><a href="bioconductor-installation.html"><i class="fa fa-check"></i><b>4</b> Bioconductor Installation</a></li>
<li class="chapter" data-level="5" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html"><i class="fa fa-check"></i><b>5</b> Amplicon Bioinformatics</a><ul>
<li class="chapter" data-level="5.1" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#libraries"><i class="fa fa-check"></i><b>5.1</b> Libraries</a></li>
<li class="chapter" data-level="5.2" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#dataset"><i class="fa fa-check"></i><b>5.2</b> Dataset</a></li>
<li class="chapter" data-level="5.3" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#trimming-filtering"><i class="fa fa-check"></i><b>5.3</b> Trimming &amp; Filtering</a></li>
<li class="chapter" data-level="5.4" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#infer-sequence-variants"><i class="fa fa-check"></i><b>5.4</b> Infer Sequence Variants</a></li>
<li class="chapter" data-level="5.5" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#merge-forward-reverse-reads"><i class="fa fa-check"></i><b>5.5</b> Merge Forward &amp; Reverse Reads</a></li>
<li class="chapter" data-level="5.6" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#construct-sequence-table-remove-chimeras"><i class="fa fa-check"></i><b>5.6</b> Construct Sequence Table &amp; Remove Chimeras</a></li>
<li class="chapter" data-level="5.7" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#assign-taxonomy"><i class="fa fa-check"></i><b>5.7</b> Assign Taxonomy</a></li>
<li class="chapter" data-level="5.8" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#construct-phylogenetic-tree"><i class="fa fa-check"></i><b>5.8</b> Construct Phylogenetic Tree</a></li>
<li class="chapter" data-level="5.9" data-path="amplicon-bioinformatics.html"><a href="amplicon-bioinformatics.html#create-phyloseq-object"><i class="fa fa-check"></i><b>5.9</b> Create Phyloseq Object</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phyloseq-object-processing.html"><a href="phyloseq-object-processing.html"><i class="fa fa-check"></i><b>6</b> Phyloseq Object Processing</a><ul>
<li class="chapter" data-level="6.1" data-path="phyloseq-object-processing.html"><a href="phyloseq-object-processing.html#filtering"><i class="fa fa-check"></i><b>6.1</b> Filtering</a></li>
<li class="chapter" data-level="6.2" data-path="phyloseq-object-processing.html"><a href="phyloseq-object-processing.html#agglomerate-taxa"><i class="fa fa-check"></i><b>6.2</b> Agglomerate Taxa</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="microbiome-analysis.html"><a href="microbiome-analysis.html"><i class="fa fa-check"></i><b>7</b> Microbiome Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="microbiome-analysis.html"><a href="microbiome-analysis.html#multivariate-projection"><i class="fa fa-check"></i><b>7.1</b> Multivariate Projection</a></li>
<li class="chapter" data-level="7.2" data-path="microbiome-analysis.html"><a href="microbiome-analysis.html#microbial-abundandce"><i class="fa fa-check"></i><b>7.2</b> Microbial Abundandce</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="classification-using-microbiome.html"><a href="classification-using-microbiome.html"><i class="fa fa-check"></i><b>8</b> Classification using Microbiome</a></li>
<li class="chapter" data-level="9" data-path="closure.html"><a href="closure.html"><i class="fa fa-check"></i><b>9</b> Closure</a></li>
<li class="chapter" data-level="10" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>10</b> References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Microbiome in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="classification-using-microbiome" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Classification using Microbiome</h1>
<p>The main microbiome analysis was already done in the previous chapter. But the result of previous chapter can be used further for additional analysis. In this capter we will demonstrate examples for doing supervised learning (a classification task) using microbiome data.</p>
<p>The classification task enhanced by microbiome data relies on the sample we have, which was obtained based on a research question. Our research question provide a design where we can obtain sample from early or old aged mouse. After microbiome analysis, we knew that microbiome composition change with age. Therefore, we can try <strong>classify whether a mouse is early or old aged based on the its gut microbioal composition</strong>.</p>
<p>The package <code>caret</code> provides various functions for machine learning algorithm. We’ll try using the robust <strong>Random Forest</strong> algorithm on this one. For that we also need a <code>randomForest</code> package. Make sure that you have each of them installed and loaded into your session.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="classification-using-microbiome.html#cb108-1"></a><span class="co"># load library</span></span>
<span id="cb108-2"><a href="classification-using-microbiome.html#cb108-2"></a><span class="kw">library</span>(caret)</span>
<span id="cb108-3"><a href="classification-using-microbiome.html#cb108-3"></a><span class="kw">library</span>(randomForest)</span></code></pre></div>
<p>We’ll be using microbial count or <code>otu_table</code> from <code>ps_cut</code> data which represent the microbial composition of our sample. This data is also already removed for its outlier. We’ll also log-transform our data before analysis.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="classification-using-microbiome.html#cb109-1"></a>ps_cut<span class="op">@</span>otu_table[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## Loading required package: phyloseq</code></pre>
<pre><code>## OTU Table:          [3 taxa and 3 samples]
##                      taxa are columns
##        ASV5 ASV8 ASV11
## F3D0    173  185   111
## F3D1    140  191   284
## F3D141  190  324   225</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="classification-using-microbiome.html#cb112-1"></a>ps_cut_log &lt;-<span class="st"> </span><span class="kw">transform_sample_counts</span>(ps_cut, <span class="cf">function</span>(x) <span class="kw">log</span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>x))</span></code></pre></div>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="classification-using-microbiome.html#cb113-1"></a>ps_cut_log<span class="op">@</span>otu_table[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## OTU Table:          [3 taxa and 3 samples]
##                      taxa are columns
##            ASV5     ASV8    ASV11
## F3D0   5.159055 5.225747 4.718499
## F3D1   4.948760 5.257495 5.652489
## F3D141 5.252273 5.783825 5.420535</code></pre>
<p>The first step of classification (or any machine learning task) is to divide our data into <strong>train</strong> and <strong>test</strong> set. to ensure that the test set realistically simulates the collection of new data, we will be using random sampling to split the data.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="classification-using-microbiome.html#cb115-1"></a><span class="co"># get age label &amp; microbiome data</span></span>
<span id="cb115-2"><a href="classification-using-microbiome.html#cb115-2"></a>data_ml &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">age =</span> <span class="kw">sample_data</span>(ps_cut_log)<span class="op">$</span>When, <span class="kw">otu_table</span>(ps_cut_log))</span>
<span id="cb115-3"><a href="classification-using-microbiome.html#cb115-3"></a></span>
<span id="cb115-4"><a href="classification-using-microbiome.html#cb115-4"></a><span class="co"># get index for sampling</span></span>
<span id="cb115-5"><a href="classification-using-microbiome.html#cb115-5"></a>idx_train &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">nrow</span>(<span class="kw">sample_data</span>(ps_cut_log)), <span class="dt">size =</span> <span class="kw">nrow</span>(<span class="kw">sample_data</span>(ps_cut_log))<span class="op">*</span><span class="fl">0.8</span>)</span>
<span id="cb115-6"><a href="classification-using-microbiome.html#cb115-6"></a></span>
<span id="cb115-7"><a href="classification-using-microbiome.html#cb115-7"></a><span class="co"># splitting train-test</span></span>
<span id="cb115-8"><a href="classification-using-microbiome.html#cb115-8"></a>training &lt;-<span class="st"> </span>data_ml[idx_train,]</span>
<span id="cb115-9"><a href="classification-using-microbiome.html#cb115-9"></a>testing &lt;-<span class="st"> </span>data_ml[<span class="op">-</span>idx_train,]</span></code></pre></div>
<p>Once we split the data, we can use the train function to fit the <strong>Random Forest</strong> model.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="classification-using-microbiome.html#cb116-1"></a>rfFit &lt;-<span class="st"> </span><span class="kw">train</span>(age <span class="op">~</span><span class="st"> </span>., </span>
<span id="cb116-2"><a href="classification-using-microbiome.html#cb116-2"></a>               <span class="dt">data =</span> training,</span>
<span id="cb116-3"><a href="classification-using-microbiome.html#cb116-3"></a>               <span class="dt">method =</span> <span class="st">&quot;rf&quot;</span>)</span>
<span id="cb116-4"><a href="classification-using-microbiome.html#cb116-4"></a></span>
<span id="cb116-5"><a href="classification-using-microbiome.html#cb116-5"></a>rfFit</span></code></pre></div>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="classification-using-microbiome.html#cb117-1"></a><span class="kw">table</span>(rfClasses, testing<span class="op">$</span>age)</span></code></pre></div>
<pre><code>##          
## rfClasses Early Late
##     Early     2    0
##     Late      0    2</code></pre>
<p>Next we can predict mouse age labels on the test set using <code>predict()</code> function.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="classification-using-microbiome.html#cb119-1"></a>rfClasses &lt;-<span class="st"> </span><span class="kw">predict</span>(rfFit, </span>
<span id="cb119-2"><a href="classification-using-microbiome.html#cb119-2"></a>                     <span class="dt">newdata =</span> testing)</span></code></pre></div>
<p>Then we can do a quick model evaluation by using simple confusion matrix.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="classification-using-microbiome.html#cb120-1"></a><span class="co"># other options use: `confusionMatrix()`</span></span>
<span id="cb120-2"><a href="classification-using-microbiome.html#cb120-2"></a><span class="kw">table</span>(<span class="dt">pred =</span> rfClasses, </span>
<span id="cb120-3"><a href="classification-using-microbiome.html#cb120-3"></a>      <span class="dt">actual =</span> testing<span class="op">$</span>age)</span></code></pre></div>
<pre><code>##        actual
## pred    Early Late
##   Early     2    0
##   Late      0    2</code></pre>
<p>We’ll there is accuracy 100%, but please note that the number of data we have for this demonstration is still very low. It might be better to train and test the model with a larger sample size, for example 30-300 sample or more.</p>
<p>To better understand the fitted random forest model, we can identify which microbe have the highest influence in the random forest prediction. We can use the function <code>importance()</code> from <code>randomForest</code> package and to find out each variable or in our case microbial species importance:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="classification-using-microbiome.html#cb122-1"></a><span class="kw">importance</span>(rfFit<span class="op">$</span>finalModel)</span></code></pre></div>
<pre><code>##        MeanDecreaseGini
## ASV5        0.125397702
## ASV8        0.219382928
## ASV11       0.163005528
## ASV15       0.271417316
## ASV16       0.333647808
## ASV27       0.405051571
## ASV30       0.239835786
## ASV34       0.000000000
## ASV41       0.148721834
## ASV44       0.014960373
## ASV45       0.141743723
## ASV46       0.165285570
## ASV47       0.000000000
## ASV48       0.147208791
## ASV52       0.255870751
## ASV53       0.000000000
## ASV58       0.242872006
## ASV60       0.000000000
## ASV65       0.086742524
## ASV66       0.280912821
## ASV68       0.301280131
## ASV71       0.117645455
## ASV74       0.306883006
## ASV81       0.175225974
## ASV86       0.069356466
## ASV89       0.106354057
## ASV92       0.000000000
## ASV94       0.131035786
## ASV99       0.268397036
## ASV101      0.000000000
## ASV102      0.087275169
## ASV111      0.016042025
## ASV120      0.000000000
## ASV121      0.163664924
## ASV132      0.221767621
## ASV141      0.240852081
## ASV143      0.103141114
## ASV145      0.000000000
## ASV146      0.058373826
## ASV164      0.098591275
## ASV172      0.070926607
## ASV177      0.068640637
## ASV190      0.024858841
## ASV191      0.008968032
## ASV203      0.063437085
## ASV208      0.026504762
## ASV217      0.010176623
## ASV224      0.023548452
## ASV233      0.029568909</code></pre>
<p>The function above shown all the variable importance, but we only need one microbe with the highest importance. In this case “ASV27”. Let’s use some function to simply take out that microbe and find out its taxonomy so we can analyze it further.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="classification-using-microbiome.html#cb124-1"></a><span class="co"># obtain taxonomy of all microbial sample</span></span>
<span id="cb124-2"><a href="classification-using-microbiome.html#cb124-2"></a>tax_rf &lt;-<span class="st"> </span><span class="kw">tax_table</span>(ps_cut_log)</span></code></pre></div>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="classification-using-microbiome.html#cb125-1"></a><span class="co"># filter microbes with the highest importance</span></span>
<span id="cb125-2"><a href="classification-using-microbiome.html#cb125-2"></a>tax_rf[<span class="kw">which.max</span>(<span class="kw">importance</span>(rfFit<span class="op">$</span>finalModel)),]</span></code></pre></div>
<pre><code>## Taxonomy Table:     [1 taxa by 7 taxonomic ranks]:
##       Kingdom    Phylum       Class        Order             Family            
## ASV27 &quot;Bacteria&quot; &quot;Firmicutes&quot; &quot;Clostridia&quot; &quot;Oscillospirales&quot; &quot;Oscillospiraceae&quot;
##       Genus           Species
## ASV27 &quot;Oscillibacter&quot; NA</code></pre>
<p>This turns out to be a microbe from Orer <em>Oscillospirales</em>, Family <em>Oscillospiraceae</em>, and genus <em>Oscillibacter</em>. Let’s try plots its abundance across samples.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="classification-using-microbiome.html#cb127-1"></a><span class="co"># get microbial abundance</span></span>
<span id="cb127-2"><a href="classification-using-microbiome.html#cb127-2"></a>imp_abd &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">otu_table</span>(ps_cut_log)[,<span class="st">&quot;ASV27&quot;</span>])</span>
<span id="cb127-3"><a href="classification-using-microbiome.html#cb127-3"></a></span>
<span id="cb127-4"><a href="classification-using-microbiome.html#cb127-4"></a><span class="co"># combine with sample data</span></span>
<span id="cb127-5"><a href="classification-using-microbiome.html#cb127-5"></a>imp_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">sample_data</span>(ps_cut_log),</span>
<span id="cb127-6"><a href="classification-using-microbiome.html#cb127-6"></a>                     <span class="dt">abund =</span> imp_abd)</span>
<span id="cb127-7"><a href="classification-using-microbiome.html#cb127-7"></a></span>
<span id="cb127-8"><a href="classification-using-microbiome.html#cb127-8"></a><span class="co"># plotting</span></span>
<span id="cb127-9"><a href="classification-using-microbiome.html#cb127-9"></a>imp_plot&lt;-<span class="st"> </span><span class="kw">ggplot</span>(imp_df, <span class="kw">aes</span>(<span class="dt">x =</span> abund)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb127-10"><a href="classification-using-microbiome.html#cb127-10"></a><span class="st">  </span><span class="kw">geom_density</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> When),</span>
<span id="cb127-11"><a href="classification-using-microbiome.html#cb127-11"></a>              <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></span>
<span id="cb127-12"><a href="classification-using-microbiome.html#cb127-12"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Abundance of Discriminative Species&quot;</span>,</span>
<span id="cb127-13"><a href="classification-using-microbiome.html#cb127-13"></a>     <span class="dt">subtitle =</span> <span class="st">&quot;Oscillibacter sp.&quot;</span>,</span>
<span id="cb127-14"><a href="classification-using-microbiome.html#cb127-14"></a>     <span class="dt">x =</span> <span class="st">&quot;Abundance&quot;</span>,</span>
<span id="cb127-15"><a href="classification-using-microbiome.html#cb127-15"></a>     <span class="dt">y =</span> <span class="st">&quot;Density of samples&quot;</span>,</span>
<span id="cb127-16"><a href="classification-using-microbiome.html#cb127-16"></a>     <span class="dt">fill =</span> <span class="st">&quot;Age&quot;</span>) <span class="op">+</span></span>
<span id="cb127-17"><a href="classification-using-microbiome.html#cb127-17"></a><span class="st">  </span><span class="co"># below is for aesthetics</span></span>
<span id="cb127-18"><a href="classification-using-microbiome.html#cb127-18"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span></code></pre></div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="classification-using-microbiome.html#cb128-1"></a>imp_plot</span></code></pre></div>
<p><img src="07-class_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>We can see that <em>Oscillibacter sp.</em> has lower abundance in the mouse early age (0-9 days) and much higher in its late age (141-150). Based on this research, <em>Oscillibacter sp.</em> may become a biomarker for mouse age for classification task.</p>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="microbiome-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="closure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
